src/api/cubrid_log.c

```diff
@@ -0,0 +1,1533 @@
+	{
+	  rc = CUBRID_LOG_SUCCESS_WITH_NO_LOGITEM;
+	}
+      else if (reply_code == ER_CDC_EXTRACTION_TIMEOUT)
+	{
+	  CUBRID_LOG_ERROR_HANDLING (CUBRID_LOG_EXTRACTION_TIMEOUT, trace_errbuf);
+	}
+    }
+
+  ptr = or_unpack_log_lsa (ptr, next_lsa);
+
+  ptr = or_unpack_int (ptr, num_infos);
+  or_unpack_int (ptr, total_length);
+
+  if (recv_data != NULL && recv_data != reply)
+    {
+      free_and_init (recv_data);
+    }
+
+  if (rc == CUBRID_LOG_SUCCESS_WITH_NO_LOGITEM)
+    {
+      goto cubrid_log_end;
+    }
+
+  if (g_log_infos_size < *total_length)
+    {
+      g_log_infos = (char *) realloc ((void *) g_log_infos, *total_length + MAX_ALIGNMENT);
+      if (g_log_infos == NULL)
+	{
+	  CUBRID_LOG_ERROR_HANDLING (CUBRID_LOG_FAILED_MALLOC, trace_errbuf);
+	}
+
+      g_log_infos_size = *total_length;
+    }
+
+  reply = PTR_ALIGN (g_log_infos, MAX_ALIGNMENT);
+  reply_size = *total_length;
+
+
+  if (css_send_request_with_data_buffer
+      (g_conn_entry, NET_SERVER_CDC_GET_LOGITEM, &rid, NULL, 0, reply, reply_size) != NO_ERRORS)
+    {
+      CUBRID_LOG_ERROR_HANDLING (CUBRID_LOG_FAILED_CONNECT, trace_errbuf);
+    }
+
+  /* extraction timeout will be modified when it is defined */
+  if (css_receive_data (g_conn_entry, rid, &recv_data, &recv_data_size, g_extraction_timeout) != NO_ERRORS)
+    {
+      CUBRID_LOG_ERROR_HANDLING (CUBRID_LOG_FAILED_CONNECT, trace_errbuf);
+    }
+
+  if (recv_data == NULL || recv_data_size != *total_length)
+    {
+      CUBRID_LOG_ERROR_HANDLING (CUBRID_LOG_FAILED_EXTRACT, trace_errbuf);
+    }
+
+  return CUBRID_LOG_SUCCESS;
+
+cubrid_log_end:
+
+  return rc;
+
+cubrid_log_error:
+
+  if (recv_data != NULL && recv_data != reply)
+    {
+      free_and_init (recv_data);
+    }
+
+  queue_entry = css_find_queue_entry (g_conn_entry->buffer_queue, rid);
+  if (queue_entry != NULL)
+    {
+      queue_entry->buffer = NULL;
+      css_queue_remove_header_entry_ptr (&g_conn_entry->buffer_queue, queue_entry);
+    }
+
+  return err_code;
+}
+
+inline static int
+cubrid_log_make_ddl (char **data_info, DDL * ddl)
+{
+  char *ptr;
+
+  char trace_errbuf[1024];
+
+  ptr = *data_info;
+
+  ptr = or_unpack_int (ptr, &ddl->ddl_type);
+  ptr = or_unpack_int (ptr, &ddl->object_type);
+  ptr = or_unpack_int64 (ptr, (INT64 *) & ddl->oid);
+  ptr = or_unpack_int64 (ptr, (INT64 *) & ddl->classoid);
+  ptr = or_unpack_int (ptr, &ddl->statement_length);
+  ptr = or_unpack_string_nocopy (ptr, &ddl->statement);
+
+  *data_info = ptr;
+
+  return CUBRID_LOG_SUCCESS;
+}
+
+inline static int
+cubrid_log_make_dml (char **data_info, DML * dml)
+{
+  char *ptr;
+
+  int i, pack_func_code;
+  int err_code;
+
+  char trace_errbuf[1024];
+
+  ptr = *data_info;
+
+  ptr = or_unpack_int (ptr, &dml->dml_type);
+  ptr = or_unpack_int64 (ptr, (INT64 *) & dml->classoid);
+  ptr = or_unpack_int (ptr, &dml->num_changed_column);
+
+  if (dml->num_changed_column)
+    {
+      // now, just malloc for validation and later, optimize it.
```

Comment:
I think it is unnecessary to malloc/free every time. So, I request you to make a simple data buffer. (after releasing the second test version)